src/databricks/labs/ucx/azure/access.py

```diff
@@ -355,25 +492,50 @@ def _apply_storage_permission(
             logger.debug(f"{role_name} permission applied for spn {principal_id} to storage account {storage.name}")
 
-    def _create_scope(self, uber_principal: PrincipalSecret, inventory_database: str):
-        logger.info(f"Creating secret scope {inventory_database}.")
+    def _create_and_get_secret_for_uber_principal(
+        self,
+        principal_secret: PrincipalSecret,
+        scope: str,
+    ) -> GetSecretResponse:
+        """Create and get a workspace secret for the principal.
+
+        If the secret scope does not, it wil be recreated. If the secret already exists, it will be overwritten.
+        """
+        logger.info(f"Creating secret scope {scope}.")
         try:
-            self._ws.secrets.create_scope(inventory_database)
+            self._ws.secrets.create_scope(scope)
         except ResourceAlreadyExists:
-            logger.warning(f"Secret scope {inventory_database} already exists, using the same")
-        self._ws.secrets.put_secret(inventory_database, "uber_principal_secret", string_value=uber_principal.secret)
+            logger.warning(f"Secret scope {scope} already exists, using the same")
+        self._ws.secrets.put_secret(scope, self._UBER_PRINCIPAL_SECRET_KEY, string_value=principal_secret.secret)
+        return self._get_secret(scope, self._UBER_PRINCIPAL_SECRET_KEY)
+
+    @retried(on=[ResourceDoesNotExist], timeout=timedelta(minutes=1))
+    def _get_secret(self, scope: str, secret: str) -> GetSecretResponse:
+        return self._ws.secrets.get_secret(scope, secret)
+
+    def _safe_delete_scope(self, scope: str) -> None:
+        try:
+            self._ws.secrets.delete_scope(scope)
+        except ResourceDoesNotExist:
+            logger.warning(f"Secret scope {scope} does not exist, skipping delete.")
+        except PermissionDenied:
+            logger.error(f"Missing permissions to delete secret scope: {scope}", exc_info=True)
 
     def load(self):
         return self._installation.load(list[StoragePermissionMapping], filename=self.FILENAME)
 
-    def _get_storage_accounts(self) -> list[str]:
+    def _get_storage_accounts(self) -> list[StorageAccount]:
         external_locations = self._locations.snapshot()
-        storage_accounts = []
+        used_storage_accounts = []
         for location in external_locations:
             if location.location.startswith("abfss://"):
                 start = location.location.index("@")
                 end = location.location.index(".dfs.core.windows.net")
                 storage_acct = location.location[start + 1 : end]
-                if storage_acct not in storage_accounts:
-                    storage_accounts.append(storage_acct)
+                if storage_acct not in used_storage_accounts:
+                    used_storage_accounts.append(storage_acct)
+        storage_accounts = []
+        for storage_account in self._azurerm.storage_accounts():
```

Comment:
Added this (small) refactor because the complexity in `_delete_uber_principal` was too high with this (duplicated) logic